File: frontend/src/runtime/dataTable.ts
/**
 * Unified tabular result for SQL, static, and future cache providers.
 * Smart Grid and future chart components should depend on this shape only.
 */

export interface DataTableColumn {
  field: string;
  label: string;
  type?: string;
  width?: number;
  align?: 'left' | 'center' | 'right';
  format?: string;
  tone?: string;
  [key: string]: unknown;
}

export interface DataTable {
  columns: DataTableColumn[];
  rows: Array<Record<string, unknown>>;
  total?: number;
  metadata?: Record<string, unknown>;
}

export function emptyDataTable(): DataTable {
  return { columns: [], rows: [], total: 0, metadata: {} };
}

export function normalizeDataTable(raw: unknown): DataTable {
  if (!raw || typeof raw !== 'object') {
    return emptyDataTable();
  }
  const data = raw as Record<string, unknown>;
  const columns = Array.isArray(data.columns)
    ? (data.columns as Array<Record<string, unknown>>).map((col) => {
        const field = String(col.field ?? col.name ?? '');
        return {
          ...col,
          field,
          label: String(col.label ?? field),
          type: col.type != null ? String(col.type) : 'string',
        } as DataTableColumn;
      })
    : [];
  const rows = Array.isArray(data.rows)
    ? (data.rows as Array<Record<string, unknown>>)
    : [];
  const total =
    typeof data.total === 'number'
      ? data.total
      : data.total != null
        ? Number(data.total)
        : rows.length;
  const metadata =
    data.metadata && typeof data.metadata === 'object'
      ? (data.metadata as Record<string, unknown>)
      : {};
  return { columns, rows, total: Number.isFinite(total) ? total : rows.length, metadata };
}

File: backend/src/main/java/com/example/lowcode/datatable/DataTable.java
package com.example.lowcode.datatable;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Unified tabular result for SQL, static, and future cache data sources.
 * Smart Grid and later chart components consume this shape only.
 */
public final class DataTable {
    private final List<Map<String, Object>> columns;
    private final List<Map<String, Object>> rows;
    private final Long total;
    private final Map<String, Object> metadata;

    public DataTable(
            List<Map<String, Object>> columns,
            List<Map<String, Object>> rows,
            Long total,
            Map<String, Object> metadata
    ) {
        this.columns = columns == null ? List.of() : List.copyOf(columns);
        this.rows = rows == null ? List.of() : List.copyOf(rows);
        this.total = total;
        this.metadata = metadata == null
                ? Map.of()
                : Collections.unmodifiableMap(new LinkedHashMap<>(metadata));
    }

    public static DataTable empty() {
        return new DataTable(List.of(), List.of(), 0L, Map.of());
    }

    public static DataTable of(
            List<Map<String, Object>> columns,
            List<Map<String, Object>> rows,
            Long total
    ) {
        return new DataTable(columns, rows, total, Map.of());
    }

    @SuppressWarnings("unchecked")
    public static DataTable fromMap(Map<String, Object> map) {
        if (map == null) {
            return empty();
        }
        List<Map<String, Object>> columns = map.get("columns") instanceof List
                ? (List<Map<String, Object>>) map.get("columns")
                : List.of();
        List<Map<String, Object>> rows = map.get("rows") instanceof List
                ? (List<Map<String, Object>>) map.get("rows")
                : List.of();
        Long total = null;
        Object totalRaw = map.get("total");
        if (totalRaw instanceof Number number) {
            total = number.longValue();
        }
        Map<String, Object> metadata = map.get("metadata") instanceof Map
                ? (Map<String, Object>) map.get("metadata")
                : Map.of();
        return new DataTable(columns, rows, total, metadata);
    }

    public List<Map<String, Object>> getColumns() {
        return columns;
    }

    public List<Map<String, Object>> getRows() {
        return rows;
    }

    public Long getTotal() {
        return total;
    }

    public Map<String, Object> getMetadata() {
        return metadata;
    }

    /**
     * Wire-compatible map for existing REST clients (columns/rows/total + optional metadata).
     */
    public Map<String, Object> toMap() {
        Map<String, Object> map = new LinkedHashMap<>();
        map.put("columns", new ArrayList<>(columns));
        map.put("rows", new ArrayList<>(rows));
        if (total != null) {
            map.put("total", total);
        }
        if (!metadata.isEmpty()) {
            map.put("metadata", new LinkedHashMap<>(metadata));
        }
        return map;
    }

    public DataTable withMetadata(Map<String, Object> extra) {
        if (extra == null || extra.isEmpty()) {
            return this;
        }
        Map<String, Object> merged = new LinkedHashMap<>(this.metadata);
        merged.putAll(extra);
        return new DataTable(columns, rows, total, merged);
    }
}
